### PR TITLE
Switch to user swappable mda-engines

### PR DESCRIPTION
**Pull request by ianhi**
As discussed in: https://github.com/tlambert03/pymmcore-plus/issues/101
Not polished yet - Opening for feedback from @tlambert03 

The big thing I struggled with was the relationship between core and the mda object should be. The mda engine needs to have access to core methods in order to function but then you can end up with this awful thing: `core._mda_engine._mmc._mda_engine._mmc.....`

If we decide to got this way then this should also get:
1. update the client to reflect the new structure.
2. Start of an `mda_tests` file
3. add an example showing how to define your own MDA engine.

### CODE REVIEW

**Comment by ianhi** [comment]
From a user perspective the main (only?) change will be that you now type `core.mda` rather then `core.` when accessing mda related things.

File: pymmcore_plus/core/_mmcore_plus.py
Comment (ianhi) [inline pymmcore_plus/core/_mmcore_plus.py:464]: added so that downstream things like napari-micromanager can unregister their callbacks with the old engine and attach to the new one.

File: pymmcore_plus/core/_mmcore_plus.py
Comment (ianhi) [inline pymmcore_plus/core/_mmcore_plus.py:461]: should have a test that this fails as expected. Also better error text

**Comment by ianhi** [comment]
image
nice :/  they all passed locally I swear!

**Comment by ianhi** [comment]
ah this is actually a bit complicated to add to remote (or at least I'm not sure exactly how to do) So going to put that off until we solidify how the local variant should work.

File: pymmcore_plus/mda/_engine.py
Comment (tlambert03) [inline pymmcore_plus/mda/_engine.py:14]: `Protocol` also actually works as an ABC.  So, if we add some `@abstractmethod` decorators here, someone could (optionally) use it as a base class and make sure their protocol is in spec.  It still supports structural sub-typing though, so they don't have to use.

So this could look like 

```python

@runtime_checkable
class PMDAEngine(Protocol):
    @abstractmethod
    def __init__(self, mmc: "CMMCorePlus") -> None:
        ...

    @abstractproperty
    def events(self) -> PMDASignaler:
        """Add docstring prescribing behavior."""

    @abstractmethod
    def cancel(self):
        """Add docstring prescribing behavior."""

    @abstractmethod
    def toggle_pause(self):
        """Add docstring prescribing behavior."""

    @abstractmethod
    def run(self, sequence: MDASequence):
        """Add docstring prescribing behavior."""
```

and then lower (optional, but nice to immediately show that it implements the protocol)

```python
class MDAEngine(PMDAEngine):
```

File: pymmcore_plus/mda/_engine.py
Comment (tlambert03) [inline pymmcore_plus/mda/_engine.py:30]: if we really wanted to, we could set self._mmc to `None`  and then have a property that gets the global instance

```
@property
def mmc(self):
    return self._mmc or CMMCorePlus.instance()
```

**Review by tlambert03 — COMMENTED** [review]
really like the direction this is going!

> The big thing I struggled with was the relationship between core and the mda object should be. The mda engine needs to have access to core methods in order to function but then you can end up with this awful thing: core._mda_engine._mmc._mda_engine._mmc.....

I think it's close to fine as it is.  This part is really an internal implementation detail right? (since the whole point is to allow for an mda object that uses some other core/engine?).  I did leave a small note about how you could make `mmc` optional in the constructor of `MDAEngine`, and then retrieve the global singleton afterwards with a property (only when needed)

Also not worried about the recursive pointers.

let me know when you feel like this is settling into finished, or if you are mostly just at the point where you're fighting with CI

File: pymmcore_plus/mda/_engine.py
Comment (ianhi) [reply pymmcore_plus/mda/_engine.py:30]: The other way I think is to call `instance` in the `run` method which is the only place it actually gets used.

File: pymmcore_plus/mda/_engine.py
Comment (ianhi) [reply pymmcore_plus/mda/_engine.py:14]: oh i re-read what you wrote. I skimmed too quickly

File: pymmcore_plus/mda/_engine.py
Comment (tlambert03) [reply pymmcore_plus/mda/_engine.py:30]: yep, that works too!

File: pymmcore_plus/mda/_engine.py
Comment (tlambert03) [reply pymmcore_plus/mda/_engine.py:14]: > I think it might also be nice to include a method like the waiting loop as a method that can be used by a subclass. Is it ok to have a method with implementation on the ABC or should that really go somewhere else?

yeah, it's technically ok... but if you're imagining something like a hook that gets called somewhere inside of `run`, then i think it actually makes the "protocol" part a lot more complicated.  Since whether that hook gets called or not really depends on whether the `run` method calls it: making it more of an implementation detail of the engine, rather than an enforceable part of the protocol

File: pymmcore_plus/mda/_engine.py
Comment (ianhi) [reply pymmcore_plus/mda/_engine.py]: or a third option of the `core.mda` being lazily created.

Comment (tlambert03) [reply pymmcore_plus/mda/_engine.py]: I think I like keeping this in mda.  I like the optional input in the `__init__`... whether you use the `@property` for `mmc`, or just use `self._mmc or CMMCorePlus.instance()` in the run method, I don't mind

**Comment by ianhi** [comment]
I'm really struggling with getting this to work through pyro. I can't seem to get past this:

```
AttributeError: remote object 'PYRO:pymmcore_plus.CMMCorePlus@127.0.0.1:54333' has no exposed attribute or method 'mda'
```

I think because the `wrap_for_pyro` isn't recognizing properties. But even if I go in an add some logic to make sure it's on the proxy class the RemoteMMCore doesn't seem to pick it up.

**Comment by ianhi** [comment]
Latest commit is my current best effort at getting remote to work.

File: pymmcore_plus/remote/_util.py
Comment (ianhi) [inline pymmcore_plus/remote/_util.py:57]: This returns true, which is why I'm confused about it not showing up on RemoteMMCore

**Comment by tlambert03** [comment]
> Any advice on getting that to work?

I say ditch it for now... skip the remote tests and make any assertions you need/want in the code to ensure that we're running locally

**Comment by tlambert03** [comment]
> AttributeError: remote object 'PYRO:pymmcore_plus.CMMCorePlus@127.0.0.1:54333' has no exposed attribute or method 'mda'

yeah, wrap_for_pyro was really quite a challenge with SWIG-compiled classes.  I asked a question about that https://github.com/irmen/Pyro5/issues/41 ... but I get the sense that the library made the assumption that it's working with pure python classes that behave normally.  A reasonable assumption, but definitely a big difficulty in our case.  

fine with me not to support that here

**Comment by ianhi** [comment]
> fine with me not to support that here

You mean that `Remote` cannot use mda? because I think that's more or less what that would amount - because you can't access the mda events.

**Comment by tlambert03** [comment]
>You mean that Remote cannot use mda?

yep... pretty much.  I really do think that until #43 is fixed, the whole thing is more or less useless anyway.  If some great need returns for it in the future, then I would 1) re-evaluate whether there is a different IPC strategy that might be better 2) fix #43 then 3) come back to fixing mda

**Comment by ianhi** [comment]
> yep... pretty much. I really do think that until https://github.com/tlambert03/pymmcore-plus/issues/43 is fixed, the whole thing is more or less useless anyway. If some great need returns for it in the future, then I would 1) re-evaluate whether there is a different IPC strategy that might be better 2) fix https://github.com/tlambert03/pymmcore-plus/issues/43 then 3) come back to fixing mda

Ok I'll drop those commits - and we can add a note in the docs about this limitation.

File: pymmcore_plus/core/_mmcore_plus.py
Comment (tlambert03) [inline pymmcore_plus/core/_mmcore_plus.py:459]: docstring

File: pymmcore_plus/core/_mmcore_plus.py
Comment (tlambert03) [inline pymmcore_plus/core/_mmcore_plus.py:454]: i know it previously didn't have one, but could you add a very brief docstring here?

File: pymmcore_plus/mda/_engine.py
Comment (tlambert03) [inline pymmcore_plus/mda/_engine.py:25]:
```suggestion
```

File: pymmcore_plus/mda/_engine.py
Comment (tlambert03) [inline pymmcore_plus/mda/_engine.py:20]:
```suggestion
```
(can remove `...` when you have a docstring.)

File: pymmcore_plus/mda/_engine.py
Comment (tlambert03) [inline pymmcore_plus/mda/_engine.py:30]:
```suggestion
```

File: pymmcore_plus/mda/_engine.py
Comment (tlambert03) [inline pymmcore_plus/mda/_engine.py:35]:
```suggestion
```

File: pymmcore_plus/mda/_engine.py
Comment (tlambert03) [inline pymmcore_plus/mda/_engine.py:40]:
```suggestion
```

**Review by tlambert03 — APPROVED** [review]
i know it's still draft, but it looks like tests are passing and the code is looking great!  Left a few minor comments, but  approving now.  Feel free to merge when ready (or ping me if you want more input)

File: pymmcore_plus/core/_mmcore_plus.py
Comment (ianhi) [reply pymmcore_plus/core/_mmcore_plus.py:454]: good call

File: pymmcore_plus/mda/_engine.py
Comment (ianhi) [inline pymmcore_plus/mda/_engine.py:55]: only real heads up is that I added this so that we can make sure an MDA isn't running when the user registers a new engine. otherwise you can end up losing track of it and have no way to cancel it.

File: pymmcore_plus/_tests/conftest.py
Comment (ianhi) [inline pymmcore_plus/_tests/conftest.py:17]: also of note - I don't think we were garunteed to get the QSignaler before. At least locally I wasn't getting it.

**Comment by ianhi** [comment]
> and we can add a note in the docs about this limitation.

Remote is actually never mentioned in the docs - so don't think it makes sense to add one for this.

> i know it's still draft, but it looks like tests are passing and the code is looking great! 
> Feel free to merge when ready

:grinning: will do in a bit

**Comment by tlambert03** [comment]
lovely!

**Comment by ianhi** [comment]
hooray! Thanks for the help talley!


Also heads up to @henrypinkard - you can make an object that conforms to the Engine protocol and then pass the actual acquisition through to micromanager

also may be useful for @ptbrown1729 to implement your own engine to use your DAQ devices. Feel free to ping me for help with that - I may end up doing that for my setup next week

**Comment by henrypinkard** [comment]
Paused, cancel, etc make sense but what is this thing:

class PMDASignaler(*args, **kwargs)

**Comment by ianhi** [comment]
It's anything that has the right names of the mda signals. e.g. this psygnal version:
https://github.com/tlambert03/pymmcore-plus/blob/main/pymmcore_plus/mda/events/_psygnal.py

which you can just import `from pymmcore_plus.mda.events import MDASignaler` and use

**Comment by henrypinkard** [comment]
It seems to me that this would be a confusing addition to the pycromanager api. Is it possible to implement a translation class within this project that translates this api to the pycromanager API?

**Comment by tlambert03** [comment]
The `PMDASignaler` is a `typing.Protocol` that declares the event/updating interface that an engine needs to provide.  We need some way for pycromanager to tell us what's happening... and that tells you the names of the signals to emit.  This goes back to the psygnal event emitter thing...

**Comment by tlambert03** [comment]
(you don't actually need to subclass that or anything, just need to provide the interface)

**Comment by henrypinkard** [comment]
I see. But this doesn't quite answer my question.

Pycromanager doesn't use events anywhere. The equivalent things happen with callback functions (when they happen at all). I am hesitant to add this to pycromanager, because it duplicates functionality with a different, and nonessential mechanism, which I think would make the API more confusing to user using pycromanager alone.

What I am suggesting is that a compatibility layer that translates the the mechanisms used by pycromanger to this event-based system live in pymmcore-plus, since this seems to be where the specification of an event-based acquisition engine lives. Does that make sense?

**Comment by tlambert03** [comment]
It does,  maybe @ianhi has an idea here, or perhaps we can zoom about this.  I understand that pycromanager doesn't use events anywhere yet, largely because all of the event-related stuff that would signal to the GUI is on the Java side where the gui was written.  I can't immediately think of how we would know that pycromanager has "done anything" without it notifying us somehow (i.e. using an Event).

It would be very much akin to what you did [here](https://github.com/micro-manager/pycro-manager/blob/10057dd5efc7167a3389383e8812e69b1513c1db/pycromanager/napari_util.py#L39-L44)

```python
                if not hasattr(napari_signaller, 'old_shape') or \
                        napari_signaller.old_shape.size != shape.size or \
                        np.any(napari_signaller.old_shape != shape):
                    image = dataset.as_array()
                    napari_signaller.old_shape = shape
                    yield image
```

except that rather than yield image, you would emit a signal `frameReady.emit(image)`.  This doesn't have to mean you're extending the public facing pycro-manager API... it's just a good way for the programs to meet half-way with loose coupling

**Comment by henrypinkard** [comment]
> largely because all of the event-related stuff that would signal to the GUI is on the Java side where the gui was written. I can't immediately think of how we would know that pycromanager has "done anything" without it notifying us somehow (i.e. using an Event).

Everything is using callbacks (on python and java sides), which do essentially the same thing. I try to studiously avoid event-based systems because I find them to be much harder to debug at compile time

> This doesn't have to mean you're extending the public facing pycro-manager API... it's just a good way for the programs to meet half-way with loose coupling

That makes sense.


Yeah maybe lets set up a zoom next week? I feel like we might be able to knock this all out in a couple hours

**Comment by tlambert03** [comment]
Call them callbacks, call them events, it's the same thing here :)

You connect the callback to the signal emitter.

**Comment by ianhi** [comment]
> What I am suggesting is that a compatibility layer that translates the the mechanisms used by pycromanger to this event-based system live in pymmcore-plus, since this seems to be where the specification of an event-based acquisition engine lives. Does that make sense?

So was I thinking this would make the most sense to have either in a new package: `pycromanager-plus-bridge` or as a sub package (like dataset) of pycromanager. So you could do:

```python
from pycromanager import Core
from pycromanager.pymmcore_plus_bridge import EngineWrapper

from pymmcore_plus import CMMCorePlus

core = Core()
engine = EngineWrapper(core)
CMMCorePlus.instance().register_mda_engine(engine)
```
and then it would be used by napari-micromanager etc. The engine wrapper would do the small work of translating pycromanager callbacks to the required signals.


As for why I think that that should live on pycromanager's side

> Yeah maybe lets set up a zoom next week? I feel like we might be able to knock this all out in a couple hours

I can zoom - I think you have my email right? We can coordinate there

**Comment by ianhi** [comment]
This new pycromanager subpackage (or freestanding package) I'm proposing could also provide a mixin class that would benefit from all the pymmcore-plus additions while still calling out to java for everything provided there. Basically this:

```python
class pycroCore:
    def somemethod(self):
        print("Calling pycromanager somemethod")


class CMMCorePlus:
    def somemethod(self):
        print("CMMCore somemethod")

    def cmmonly(self, somearg):
        """A method only on CMMCorePlus"""
        self.somemethod()
        print(f"got {somearg}")


class pycroCoreBridge(pycroCore, CMMCorePlus):
    ...


core = pycroCoreBridge()
core.somemethod()
core.cmmonly("blarg")
```

then all that would needed is some workin into pycroCore init to convert pycromanager style callbacks to CMMCorePlus style and I'm pretty sure it would just work every where that CMMCorePlus does.


Also now that I think about it more. I think a standalone package makes the most sense. It reflects the reality that it's a bridge between libraries that have more of sibling than parent child relationship. Once you put the bridge in one or the other then there's an awkward coupling that's taken place.

(I would give a more complete and tested example - but unfortunately I've never been able to use pycromanager on my laptop because I can't install full micro-manager)

**Pull request merged**
